src/media/js/apps.js

```diff
@@ -113,10 +113,59 @@ define('apps',
         if (product.payment_required && !product.price) {
             reasons.push(gettext('not available for your region'));
         }
-        if (!product.isWebsite && !capabilities.webApps ||
-            (!capabilities.packagedWebApps && product.is_packaged) ||
-            !_.contains(product.device_types, device)) {
-            reasons.push(gettext('not available for your platform'));
+        if (!capabilities.webApps &&
+            !_.contains(product.device_types, 'firefoxos') &&
+            !_.contains(product.device_types, 'android-mobile') &&
+            !_.contains(product.device_types, 'android-tablet') &&
+            _.contains(product.device_types, 'desktop')) {
+            reasons.push(gettext('try this app on Firefox Desktop'));
+        }
+        if (!capabilities.firefoxAndroid &&
+            !_.contains(product.device_types, 'desktop') &&
+            !_.contains(product.device_types, 'firefoxos') &&
+            (_.contains(product.device_types, 'android-mobile') ||
+            _.contains(product.device_types, 'android-tablet'))) {
+            reasons.push(gettext('try this app on Android'));
+        }
+        if (!capabilities.firefoxOS &&
+            !_.contains(product.device_types, 'desktop') &&
+            !_.contains(product.device_types, 'android-mobile') &&
+            !_.contains(product.device_types, 'android-tablet') &&
+            _.contains(product.device_types, 'firefoxos')) {
+            reasons.push(gettext('try this app on Firefox OS'));
```

Comment:
So, we can simplify these checks. 
1) We don't care about showing incompatibility message for Websites, IMHO.
2) We only need to keep 4 messages, we don't need to show messages like "try this app on Desktop / Android", only one suggestion for each device should be enough.

I suggest to replace all your code by the following : 
```
/* 
 * Return message to show the client when a product is not compatible, trying
 * to find a compatible device to give the user an actionable suggestion.
 */
function getIncompatibilityMessage(product) {
    var message;

    if (_.contains(product.device_types, 'firefoxos')) {
        message = gettext('try this app on Firefox OS'));
    } else if (_.contains(product.device_types, 'android-mobile')) {
        message = gettext('try this app on Firefox for Android Mobile'));
    } else if (_.contains(product.device_types, 'android-tablet')) {
        message = gettext('try this app on Firefox for Android Tablet'));
    } else if (_.contains(product.device_types, 'desktop')) {
        message = gettext('try this app on Firefox'));
    }
    return message;
}

if (!product.isWebsite) {
    // The client is missing some capabilities.
    if (!capabilities.webApps || 
        (!capabilities.packagedWebApps && product.is_packaged)) {
            reasons.push(getIncompatibilityMessage(product));
    } 
    // The product is incompatible with our client.
    else if (!_.contains(product.device_types, device)) {
        reasons.push(getIncompatibilityMessage(product));
    }
}
```

Can you make that work, adjust the unit tests ? Then we'll look at the slimerjs failures.